packages/financial-templates-lib/src/logger/Logger.js

```diff
@@ -42,11 +42,25 @@ async function waitForLogger(logger) {
 // If the log entry contains an error then extract the stack trace as the error message.
 function errorStackTracerFormatter(logEntry) {
   if (logEntry.error) {
-    logEntry.error = logEntry.error.stack;
+    logEntry.error = handleRecursiveErrorArray(logEntry.error);
   }
   return logEntry;
 }
 
+// Handle case where `error` is an array of errors and we want to display all
+// of the error stacks recursively. i.e. `error` is in the shape:
+// [[Error, Error], [Error], [Error, Error]]
+function handleRecursiveErrorArray(error) {
+  // If error has a `stack` param, then no need to recurse anymore:
+  if (error.stack) return error.stack;
+
+  let results = [];
+  error.forEach(_error => {
+    results.push(handleRecursiveErrorArray(_error));
+  });
+  return results;
```

Comment:
A few questions/comments:
1. This is really cool -- I like it a lot.
2. I get that this message is only meant to handle the error stacks, but what about the error message? Does that get extracted as well somewhere? Or is it implicitly included in the stack?
3. If I'm understanding this loop correctly, this seems to preserve the array nesting. So you end up passing back `[[stack, stack], stack, [stack, stack]]` instead of `[stack, stack, stack, stack, stack]`. Is that desired? My initial understanding was that you wanted to flatten this by doing something more like (using the spread operator):
```js
  let results = [];
  error.forEach(_error => {
    results.push(...handleRecursiveErrorArray(_error));
  });
  ```
4. This might be a slightly more readable way to structure the loop (won't work if you want to flatten as suggested in 3):
```js
return error.map(handleRecursiveErrorArray);
```

Reply:
> I get that this message is only meant to handle the error stacks, but what about the error message? Does that get extracted as well somewhere? Or is it implicitly included in the stack?

The message is part of the stack, that's why currently we just log out the stack.

> My initial understanding was that you wanted to flatten this by doing something more like (using the spread operator):
I do want to flatten it, does the `...` syntax achieve this?

> This might be a slightly more readable way to structure the loop (won't work if you want to flatten as suggested in 3):
Oo I like this but I do want to flatten it.

Reply:
As discussed IRL, when changing the line to `return error.map(...).flat()` you get this for BasketSpread (which is composed of Medianizers):

```
[warn] nick-local-testing (SyntheticPegMonitor)⭢Unable to get volatility data, missing historical price data
• error:
- 0: Error: CryptoWatch-uniswap-v2-fraxusdc: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:118:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:116:58)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)





3:15
- 1: Error: CryptoWatch-uniswap-v2-esdusdc: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:118:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:116:58)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
3:15
- 2: Error: Uniswap-0xd4405f0704621dbe9d4dea60e128e0c3b26bddbd missing historical price @ time 1612293348
   at UniswapPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js:75:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:118:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:116:58)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
3:15
- 3: Error: CryptoWatch-bitfinex-usdtusd: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:125:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:123:50)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
3:15
- 4: Error: CryptoWatch-kraken-usdtusd: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:125:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:123:50)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
3:15
- 5: Error: CryptoWatch-kraken-usdcusd: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:125:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:123:50)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
3:15
- 6: Error: CryptoWatch-bitstamp-usdcusd: no valid price periods
   at CryptoWatchPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js:85:13)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:44:26
   at Array.map (<anonymous>)
   at MedianizerPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:42:44)
   at /Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:125:26
   at Array.map (<anonymous>)
   at BasketSpreadPriceFeed.getHistoricalPrice (/Users/nicholaspai/UMA/protocol/packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:123:50)
   at SyntheticPegMonitor._calculateHistoricalVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:337:28)
   at SyntheticPegMonitor._checkPricefeedVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:293:24)
   at SyntheticPegMonitor.checkPegVolatility (/Users/nicholaspai/UMA/protocol/packages/monitors/src/SyntheticPegMonitor.js:186:28)
   at retry.retries (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:254:33)
   at processTicksAndRejections (internal/process/task_queues.js:93:5)
   at run (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:229:7)
   at Poll (/Users/nicholaspai/UMA/protocol/packages/monitors/index.js:361:5)
• lookback: 3600
```